packages/lib/src/vis/surface/hooks.ts

```diff
@@ -0,0 +1,60 @@
+import type { NumArray } from '@h5web/shared';
+import { getDims } from '@h5web/shared';
+import type { NdArray } from 'ndarray';
+import { useMemo } from 'react';
+import { Float32BufferAttribute, Uint8BufferAttribute } from 'three';
+
+export function useBufferAttributes(
+  dataArray: NdArray<NumArray>,
+  dataToColorScale: (val: number) => [number, number, number]
+) {
+  const { position, color } = useMemo(() => {
+    const { rows, cols } = getDims(dataArray);
+
+    const positions: number[] = [];
+    const colors: number[] = [];
+
+    for (let row = 0; row < rows; row++) {
+      for (let col = 0; col < cols; col++) {
+        const value = dataArray.get(row, col);
+        positions.push(col, row, value);
+        colors.push(...dataToColorScale(value));
+      }
+    }
+
+    return {
+      position: new Float32BufferAttribute(positions, 3),
+      color: new Uint8BufferAttribute(colors, 3, true),
+    };
+  }, [dataArray, dataToColorScale]);
+
+  const index = useMemo(() => {
+    const { rows, cols } = getDims(dataArray);
+
+    const indices: number[] = [];
+
+    function getIndex(col: number, row: number) {
+      return row * cols + col;
+    }
+
+    for (let row = 0; row < rows - 1; row++) {
+      for (let col = 0; col < cols - 1; col++) {
+        // Two triangles by point
+        indices.push(
+          // First triangle: (x,y) -> (x + 1, y) -> (x, y + 1)
+          getIndex(col, row),
+          getIndex(col + 1, row),
+          getIndex(col, row + 1),
+          // Second triangle: (x + 1 ,y) -> (x, y + 1) -> (x + 1, y + 1)
+          getIndex(col + 1, row),
+          getIndex(col, row + 1),
+          getIndex(col + 1, row + 1)
+        );
+      }
+    }
+
+    return indices;
+  }, [dataArray]);
```

Comment:
Difference with the `ScatterVis` hook: for `GL_POINTS`, there is only the need for a single coordinate per data points. But for `GL_TRIANGLES`, I need 6 points (2 triangles) to draw a square per data point (except the last row and column).

Rather than having 5 additional points per datapoints, I make use of [an indexed geometry](https://threejs.org/examples/#webgl_buffergeometry_indexed) to tell in which order the points must be linked to draw the triangles. This allows to reuse points and therefore to have less points in the BufferAttribute for better performance.

Also for performance, I also chose to keep the computation of `index` in a separate loop in its own `useMemo` so that they are not recomputed at each change of `colorMap`.

Reply:
:+1:  - I guess the shader will eventually be useful to avoid this double loop.